### PR TITLE
Remove path-array

### PR DESCRIPTION
**Pull request by mhart**
This reverts commits 81eadc5692eb33f308209412daa7db5d53bcb744 and ff88e5fd1211877a9a6eb2a906f7e4d4ff377604.

Of the 5.5MB footprint of node-gyp, path-array makes up 3.6MB.

Removing this reduces the weight of npm (and ultimately node) appreciably (by 20%)

### CODE REVIEW

**Comment by rvagg** [comment]
Does this actually work though? I was looking at this a while back because the dependency depth here bothers me too but I wasn't sure that the same functionality could be replicated quite as easily.

Perhaps it's really this easy but since we don't have test cases around this I'm not overly confident! We can certainly give it a try in v4 pre-releases which I'm planning on spinning up soon but tests would be nice. Any chance you'd like to write us some tests for this?

**Comment by mhart** [comment]
Sure – I mean, I believe it was only added by @TooTallNate originally because he thought it was a good fit for the `path-array` module – not because the original code, written by @timbertson was faulty in any way: https://github.com/nodejs/node-gyp/pull/509/files#r20182682

I just reverted those lines to how they were when #509 was merged.

Kinda seems like overkill to add a module for just a single string join, no?

File: lib/configure.js [hunk 1]
```diff
@@ -288,8 +287,9 @@ function configure (gyp, argv, callback) {
       argv.unshift(gyp_script)
 
       // make sure python uses files that came with this particular node package
-      var pypath = new PathArray(process.env, 'PYTHONPATH')
-      pypath.unshift(path.join(__dirname, '..', 'gyp', 'pylib'))
+      var pypath = process.env.PYTHONPATH ? [process.env.PYTHONPATH] : []
+      pypath.unshift(path.resolve(__dirname, '..', 'gyp', 'pylib'))
```
Comment (mhart) [inline lib/configure.js]: Ah, just noticed this should stay as `path.join` – will add that.

**Comment by mhart** [comment]
I'll add some tests tomorrow if I can figure out how 👍

File: lib/configure.js [hunk 2]
```diff
@@ -288,8 +287,9 @@ function configure (gyp, argv, callback) {
       argv.unshift(gyp_script)
 
       // make sure python uses files that came with this particular node package
-      var pypath = new PathArray(process.env, 'PYTHONPATH')
+      var pypath = process.env.PYTHONPATH ? [process.env.PYTHONPATH] : []
       pypath.unshift(path.join(__dirname, '..', 'gyp', 'pylib'))
+      process.env.PYTHONPATH = pypath.join(process.platform === 'win32' ? ';' : ':')
```
Comment (bnoordhuis) [inline lib/configure.js]: It's arguably a bit nicer to wrap it in a `if ('PYTHONPATH' in process.env) { ... }`.

Can you replace the `process.platform === 'win32' ? ';' : ':'` with `win ? ';' : ':'`?

**Comment by TooTallNate** [comment]
It's only so big because of `es6-symbol`... I wonder if there's a smaller polyfill.

File: lib/configure.js (hunk 2 above)
Comment (mhart) [inline lib/configure.js]: When you say wrap in an `if`, what did you have in mind?

I can certainly do:
```js
var pypath = 'PYTHONPATH' in process.env ? [process.env.PYTHONPATH] : []
```

Or would you prefer something like:

```js
var pypath = [path.join(__dirname, '..', 'gyp', 'pylib')]
if ('PYTHONPATH' in process.env) {
  pypath.push(process.env.PYTHONPATH)
}
process.env.PYTHONPATH = pypath.join(win ? ';' : ':')
```
(I kinda prefer that over an unshift anyway tbh)

Comment (bnoordhuis) [inline lib/configure.js]: I mean the second one.  I wrote `'PYTHONPATH' in process.env` but perhaps checking for truthiness / non-emptiness (i.e., just `process.env.PYTHONPATH`) is better.

Comment (mhart) [inline lib/configure.js]:
```js
var pypath = [path.join(__dirname, '..', 'gyp', 'pylib')]
if (process.env.PYTHONPATH) pypath.push(process.env.PYTHONPATH)
process.env.PYTHONPATH = pypath.join(win ? ';' : ':')
```
Preferences on indentation, braces?

Comment (bnoordhuis) [inline lib/configure.js]: node-gyp uses two spaces and (predominantly, I think) braces.

**Comment by mhart** [comment]
@rvagg added some tests 👍

**Comment by mhart** [comment]
@rvagg let me know if those tests look ok, or if there's anything you need further to get this through

**Comment by mhart** [comment]
Hi all – any movement on this?

FWIW, I've been building a tool to track down size regressions (and other issues with sub-deps) – you can see an example of it coming to life here: http://pkgsize.com.s3.amazonaws.com/node-gyp.html (hovering over should show you what events caused certain size changes)

Obviously, `path-array` and `request` are the big offenders. Considering that `path-array` is literally just used for one string join here and has such a deep dep tree, it really will reduce the risk of even further bloat to remove it.

**Comment by bnoordhuis** [comment]
LGTM.  CI: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/12/

This probably also needs a CI run with v0.10 and v0.12.

**Comment by mhart** [comment]
@bnoordhuis would love to get this in – how do I get a CI run with v0.10 and v0.12?

**Comment by bnoordhuis** [comment]
https://ci.nodejs.org/view/All/job/nodegyp-test-commit/13/ - v0.10.46
https://ci.nodejs.org/view/All/job/nodegyp-test-commit/14/ - v0.12.15

@nodejs/build Can someone take a look at why the win2012r2 bot complains about this:
```
c:\workspace\nodegyp-test-commit\nodes\win2012r2\local_node>mkdir npm-cache 

c:\workspace\nodegyp-test-commit\nodes\win2012r2\local_node>call "c:\workspace\nodegyp-test-commit\nodes\win2012r2\local_node\npm.cmd" config set cache "c:\workspace\nodegyp-test-commit\nodes\win2012r2\local_node\npm-cache" 
The system cannot execute the specified program.
```
From https://ci.nodejs.org/view/All/job/nodegyp-test-commit/13/nodes=win2012r2/console.

**Comment by bnoordhuis** [comment]
The ubuntu buildbot seems pretty happy though so I think this is good to go.  @mhart Can you squash it into what you think are logical/reasonable commits?

**Comment by mhart** [comment]
@bnoordhuis Sure thing – I'll keep the original "revert" commits separate and squash the rest

**Comment by mhart** [comment]
@bnoordhuis done – let me know if you want me to squash further

**Comment by joaocgreis** [comment]
The job was trying to download `node.exe` from the wrong place, so it was getting a html page instead of a exe, hence the "cannot execute" error. Also, npm was interfering with the CITGM job. Both fixed.

Started two jobs with similar parameters:
- v0.10: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/29/
- v0.12: https://ci.nodejs.org/view/All/job/nodegyp-test-commit/30/

Seems good!

**Comment by mhart** [comment]
Anything more need to be done on this? If so lemme know 👍

**Comment by mhart** [comment]
Hey all, just checking in on this PR. Let me know if there are any more concerns?

**Comment by bnoordhuis** [comment]
I think we were (and are) just waiting for the next node-gyp release.  I've landed it in 9c8d275...ddac348, thanks!

**Comment by mhart** [comment]
Great, thanks @bnoordhuis 👍 

Once it makes it to Node.js it'll be the first drop in size we've seen in years – here's hoping!

**Comment by jbergstroem** [comment]
This is pretty awesome. Thanks for the effort!

**Comment by mhart** [comment]
Thanks for pushing this out – npm@4.1.2 has just included it and has reaped the benefits of an 18% reduction in size 👍

**Pull request closed without merging**